stream/src/main/java/org/apache/kafka/clients/processor/internals/KStreamThread.java

```diff
@@ -0,0 +1,308 @@
+    private long recordsProcessed;
+
+    protected final ConsumerRebalanceCallback rebalanceCallback = new ConsumerRebalanceCallback() {
+        @Override
+        public void onPartitionsAssigned(Consumer<?, ?> consumer, Collection<TopicPartition> assignment) {
+            ingestor.init();
+            addPartitions(assignment);
+        }
+
+        @Override
+        public void onPartitionsRevoked(Consumer<?, ?> consumer, Collection<TopicPartition> assignment) {
+            commitAll(time.milliseconds());
+            removePartitions();
+            ingestor.clear();
+        }
+    };
+
+    @SuppressWarnings("unchecked")
+    public KStreamThread(Class<? extends PTopology> topologyClass, ProcessorProperties properties) throws Exception {
+        super();
+
+        if (properties.timestampExtractor() == null)
+            throw new NullPointerException("timestamp extractor is missing");
+
+        this.metrics = new Metrics();
+        this.config = new ProcessorConfig(properties.config());
+        this.topologyClass = topologyClass;
+
+        this.properties = properties;
+        this.streamingMetrics = new KafkaStreamingMetrics();
+
+        // build the topology without initialization to get the topics for consumer
+        PTopology topology = topologyClass.getConstructor().newInstance();
+        topology.build();
+
+        // create the producer and consumer clients
+        Producer<byte[], byte[]> producer = new KafkaProducer<>(properties.config(), new ByteArraySerializer(), new ByteArraySerializer());
+        this.collector = new RecordCollectorImpl(producer, (Serializer<Object>) properties.keySerializer(), (Serializer<Object>) properties.valueSerializer());
+
+        Consumer<byte[], byte[]> consumer = new KafkaConsumer<>(properties.config(), rebalanceCallback, new ByteArrayDeserializer(), new ByteArrayDeserializer());
+        this.ingestor = new IngestorImpl(consumer, topology.topics());
+
+        this.running = true;
+        this.lastCommit = 0;
+        this.nextStateCleaning = Long.MAX_VALUE;
+        this.recordsProcessed = 0;
+        this.time = new SystemTime();
+    }
+
+    /**
+     * Execute the stream processors
+     */
+    @Override
+    public synchronized void run() {
+        log.info("Starting a kstream thread");
+        try {
+            ingestor.open();
+            runLoop();
+        } catch (RuntimeException e) {
+            log.error("Uncaught error during processing: ", e);
+            throw e;
+        } finally {
+            shutdown();
+        }
+    }
+
+    private void shutdown() {
+        log.info("Shutting down a kstream thread");
+        commitAll(time.milliseconds());
+
+        collector.close();
+        ingestor.close();
+        removePartitions();
+        log.info("kstream thread shutdown complete");
+    }
+
+    /**
+     * Shutdown this streaming thread.
+     */
+    public synchronized void close() {
+        running = false;
```

Comment:
The `run()` and `close()` methods should not be synchronized. Because they are, then once `run()` is called it will block any other synchronized method, including `close()`, and because `run()` only completes when `close()` is called, `run()` will never complete. In other words, the thread will never stop.

You should be able to simply remove the `synchronized` keywords with the current code and maintain thread safety of the `running` volatile boolean field: the only method that reads that field is the private `stillRunning()` (called via private `runLoop()` which is called via public `run()`), while the only method that writes to the field is `close()`.